ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/ReexecutingExecutionPayloadBlockManager.java

```diff
@@ -0,0 +1,172 @@
+    return new ReexecutingExecutionPayloadBlockManager(
+        recentChainData, blockImporter, pendingBlocks, futureBlocks, validator, asyncRunner);
+  }
+
+  @Override
+  public SafeFuture<?> doStart() {
+    return super.doStart()
+        .thenRun(
+            () ->
+                cancellable =
+                    Optional.of(
+                        asyncRunner.runWithFixedDelay(
+                            this::execute,
+                            RECHECK_INTERVAL,
+                            error -> LOG.error("Failed to reexecute execution payloads", error))));
+  }
+
+  @Override
+  protected SafeFuture<?> doStop() {
+    return super.doStop()
+        .thenRun(
+            () -> {
+              cancellable.ifPresent(Cancellable::cancel);
+              cancellable = Optional.empty();
+            });
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    super.onSlot(slot);
+    pendingBlocks.removeIf(block -> block.getSlot().isLessThan(slot.plus(EXPIRES_IN_SLOT)));
+  }
+
+  @Override
+  public SafeFuture<BlockImportResult> importBlock(SignedBeaconBlock block) {
+    return super.importBlock(block)
+        .thenPeek(
+            blockImportResult -> {
+              if (requiresReexecution(blockImportResult)) {
+                pendingBlocks.add(block);
+              }
+            });
+  }
+
+  private void execute() {
+    if (reexecutingExecutionPayload.compareAndSet(false, true)) {
+      LOG.debug("re-executing execution payloads for queued blocks");
+      SafeFuture.allOfFailFast(
+              pendingBlocks.stream()
+                  .map(this::reexecuteExecutionPayload)
+                  .toArray(SafeFuture[]::new))
+          .thenPeek(__ -> reexecutingExecutionPayload.set(false))
+          .exceptionally(
+              error -> {
+                if (error instanceof StopExecutionError) {
+                  LOG.error(
+                      "An error occurred which prevents additional payload executions: {} reason: {}",
+                      error.getMessage(),
+                      error.getCause().getMessage());
+                  return null;
+                }
+                throw new RuntimeException(error);
+              })
+          .reportExceptions();
+    }
+  }
+
+  private SafeFuture<Void> reexecuteExecutionPayload(final SignedBeaconBlock block) {
+    return super.importBlock(block)
+        .thenAccept(
+            blockImportResult -> {
+              if (canBeRemoved(blockImportResult)) {
+                pendingBlocks.remove(block);
+              }
+              applySubsequentExecutionPreventionCheck(blockImportResult);
+            });
+  }
+
+  private boolean requiresReexecution(BlockImportResult blockImportResult) {
+    final FailureReason failureReason = blockImportResult.getFailureReason();
+    return failureReason == FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION
+        || failureReason == FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION_SYNCING;
```

Comment:
I wonder if this should move to a method on `FailureReason` and then reuse that method in `canBeRemoved` and `applySubsequentExecutionPreventionCheck` below.